integration/experiment/ffnet/gen_region_parameters.py

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python3
+#  Copyright (c) 2015 - 2024, Intel Corporation
+#  SPDX-License-Identifier: BSD-3-Clause
+#
+
+import json
+import pandas as pd
+from sklearn import datasets, linear_model
+import argparse
+
+def get_domains(table_stats):
+    domains = []
+    for domain in ['cpu', 'gpu', 'uncore']:
+        if f'{domain}-frequency' in table_stats.columns:
+            domains.append(domain)
+    return domains
+
+def get_domain_freq_range(domain, table_stats):
+    if f'{domain}-frequency' in table_stats:
+        return {'min_freq' : min(table_stats[f'{domain}-frequency']),
+                'max_freq' : max(table_stats[f'{domain}-frequency'])}
+
+#Outputs runtime = slope * inv_freq + intercept
+def per_region_regression(table_stats, domain='cpu'):
+
+    regr = linear_model.LinearRegression()
+    region_regression = {}
+
+    for region in table_stats['app-config'].unique():
+        region_inv_freq = 1.0 / table_stats[table_stats['app-config'] == region][f'{domain}-frequency']
+        region_runtime = table_stats[table_stats['app-config'] == region]['runtime (s)']
+
+        region_inv_freq = region_inv_freq.values.reshape(-1, 1)
+        region_runtime = region_runtime.values.reshape(-1, 1)
+
+        regr.fit(region_inv_freq, region_runtime)
+
+        region_regression[region] = {"slope":regr.coef_, "intercept":regr.intercept_}
+
+    return region_regression
+
+def get_best_runtime_freq(region_regression, domain, region_name, perf_deg_factor, freq_range):
+
+    freq_max = freq_range['max_freq']
+    freq_min = freq_range['min_freq']
+    # use the region regression to find min freq under perf deg allowance
+
+    best_runtime = region_regression[region_name]["slope"] / freq_max + region_regression[region_name]["intercept"]
+
+    runtime_freq = region_regression[region_name]["slope"] / (best_runtime*perf_deg_factor - region_regression[region_name]["intercept"])
+
+    runtime_freq = min(max(runtime_freq, freq_min), freq_max)
+    return (float)(runtime_freq), (float)(best_runtime*perf_deg_factor)
+
+def get_energy_at_freq(table_stats, region, domain, freq):
+    if domain == 'gpu':
+        energy_col = 'gpu-energy (J)'
+    else:
+        energy_col = 'package-energy (J)'
+
+    freq_subset = table_stats[table_stats['app-config'] == region]
+
+    # Approximate by the nearest frequency
+    freq = freq_subset.iloc[(freq_subset[f'{domain}-frequency']-freq).abs().argsort()[:1]][f'{domain}-frequency'].tolist()[0]
```

Comment:
I think this is doing:
```python
freq = freq_subset.iloc[
    (freq_subset[f'{domain}-frequency']-freq).abs()  # distance between query and LUT freqs
    .argsort()[:1]  # argsort the distances, keeping a Series of only the first (argmin) element
][f'{domain}-frequency'].tolist()[0]  # get the frequencies at the list-of-argmin indices, extract the single value
```

Alternatively, I think something along the lines of this might be more direct:
```python
freq = freq_subset.iloc[
    (freq_subset[f'{domain}-frequency']-freq).abs().argmin()  # Returns one element instead of a list
][f'{domain}-frequency']
```